docs/proposals/service-design.md

```diff
@@ -0,0 +1,86 @@
+
+## Motivation
+The cloud native and microservice architecture becoming more and more popular and the edge node is becoming more and more powerful. The user want to separate their application to multiple microservice and some of them need to be deployed in edge node to get close to the data generated by deivces. By simply the development, the user need to use the same michanism to do the service discovery and communication with what they are doing on cloud.
+
+## Constraints and Assumptions
+1. Since the edge environment can be some non-standard system to running on some specific devices so we should not rely on the kernel network capability. In that way, we need to use userspace proxy to implement the data proxy on edge.
+2. With the same reason, clusterIP is not supported on Edge.
+3. With the same reason, no DNS services available for edge.
+4. The identification and authentication should be handled by microservices.
+5. When service communication between edge to edge, they may work offline. With the purpose to deliver a "expected result" to the user system, the service definition should be push down to the edge right after the definition operation. That is to say, the istio model may not be suitable for edge, in istio model, the service definition is pull from K8S master when the service calling happens.
+6. Only support HTTP communication in the first step
+
+## Use cases
+<img src="../images/proposals/service-design.png">
+
+### Register the service located on cloud
+* CS0.1: User create service(type ClusterIP) in K8S API Server
+* CS0.2: EC get the service definition and create router rules with the edge servicebus as source and service clusterIP as target.
+* CS0.3: Send the service definition to edge
+
+### Register the service located on edge
+* CS1: User create service(headless service without ClusterIP) in K8S API Server
+* CS2, CS3: EC get the service definiton and update the service ClusterIP to itself. Then create router rule with the EC(router) address as the source and edge servicebus as target.
+* CS4: Send the service definition to edge
+
+### Service discovery on cloud
+* Application access service with the same mechanism no matter it located on cloud or edge
+
+### Service discovery on edge
+* Need to setup http_proxy in the container(can be done automatically) with the address of ServiceBus. Then the application can vist the service provider via `http://{service_name}/resource`. ServiceBus get the service name part from the request as a reverse proxy, then pick up one POD to redirect the request.
+
+### Service communication from cloud to edge
+* ESD1: Application on cloud visit service with clusterIP and request goes to EC(refer to step CS2, CS3)
+* ESD2: Router find out the PODs in this service and pick one with some kind of load balance mechanism. Then send the request to the choosen edge with the router rule set in CS3.
+* ESD3: EdgeHub/ServiceBus get the request and redirect to the POD
```

Comment:
In my mind, cloudhub/edgehub is the channel for control plane. For data plane, we build the mesh.

Reply:
cloudhub/edgehub act as the data gateway for each node so both control and data channel data will go with them. But in those components, we have different data wrapper for the control/data packages so we do distinguish them logically.

Reply:
do we mean to say edgehub will contain two separate data wrapper modules to handle control and data packets separately ? currently edgehub only has websocket based channel for control plane right ? we plan to put service bus as a separate module parallel to edgehub right ? I am not sure how will we distinguish the packets and where will be do this.